gh-68320, gh-88302 - Allow for `pathlib.Path` subclassing

barneygale opened the pull request:
Users may wish to define subclasses of `pathlib.PurePath` and `Path` to add or modify existing methods. Before this change, attempting to instantiate a subclass raised an exception like:

    AttributeError: type object 'PPath' has no attribute '_flavour'

Previously the `_flavour` attribute was assigned as follows:

    PurePath._flavour        = xxx not set!! xxx
    PurePosixPath._flavour   = _PosixFlavour()
    PureWindowsPath._flavour = _WindowsFlavour()

It's now set as follows:

    PurePath._flavour        = os.path
    PurePosixPath._flavour   = posixpath
    PureWindowsPath._flavour = ntpath

Functionality from `_PosixFlavour` and `_WindowsFlavour` is moved into `PurePath` as underscored-prefixed classmethods. Flavour classes are removed.

A deeper dive into this patch can be read here: https://discuss.python.org/t/make-pathlib-extensible/3428/42

Fixes #68320 #88302

<!-- gh-issue-number: gh-68320 -->
* Issue: gh-68320
<!-- /gh-issue-number -->

barneygale committed:
bpo-24132, bpo-44136 - Fix `pathlib.Path` subclassing.

Users may wish to define subclasses of `pathlib.Path` to add or modify
existing methods. Before this change, attempting to instantiate a subclass
raised an exception like:

    AttributeError: type object 'PPath' has no attribute '_flavour'

Previously the `_flavour` attribute was assigned as follows:

    PurePath._flavour        = xxx not set!! xxx
    PurePosixPath._flavour   = _PosixFlavour()
    PureWindowsPath._flavour = _WindowsFlavour()

This commit replaces it with a `_pathmod` attribute, set as follows:

    PurePath._pathmod        = os.path
    PurePosixPath._pathmod   = posixpath
    PureWindowsPath._pathmod = ntpath

Functionality from `_PosixFlavour` and `_WindowsFlavour` is moved into
`PurePath` as underscored-prefixed classmethods. Flavours are removed.

barneygale committed:
Add tests

barneygale committed:
Add tests for PurePath subclasses

AlexWaygood commented on Misc/NEWS.d/next/Library/2022-03-05-02-14-09.bpo-24132.W6iORO.rst:
```diff
@@ -0,0 +1,2 @@
+Instantiating a direct subclass of :class:`pathlib.Path` raised
+:exc:`AttributeError`.
```
```suggestion
Make :class:`pathlib.Path` subclassable. Previously, attempting to subclass ``Path``
resulted in an :exc:`AttributeError`. Patch by Barney Gale.
```

merwok commented on Lib/pathlib.py:
```diff
@@ -45,240 +53,11 @@ def _is_wildcard_pattern(pat):
-class _Flavour(object):
-    """A flavour implements a particular (platform-specific) set of path
-    semantics."""
-
```
Is there a way to minimize the diff and preserve line history for this code?  For example making a base class here and inheriting from it in PurePath.

I know history is not the foremost concern, but there is often value in finding who added what line for what reason, and git tooling is still not very smart for this.

AlexWaygood replied:
@merwok, I agree that we should clearly aim for the smallest diff possible in every PR in order to avoid code churn and preserve history. But surely concerns about history should not be guiding substantive decisions about code structure, such as e.g. influencing decisions around a class's mro?

merwok replied:
I guess that ship has sailed with the removal of accessors and much code being moved!

barneygale replied:
I'm a big `git blame` fan and I try to avoid moving code around if I can help it! But I think adding another class to the MRO for the sake of a tidy history might be a bridge too far in this case.

One thing I could do to help reduce the diff is to rename `_pathmod` to `_flavour`?

barneygale committed:
Update Misc/NEWS.d/next/Library/2022-03-05-02-14-09.bpo-24132.W6iORO.rst

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

barneygale committed:
Rename `PurePath._pathmod` to `PurePath._flavour` to reduce the diff.

barneygale committed:
Merge branch 'bpo-44136-remove-pathlib-flavour-2' of github.com:barneygale/cpython into bpo-44136-remove-pathlib-flavour-2

barneygale committed:
Undo an unnecessary change to `Path.__new__()`

barneygale committed:
Remove `_casefold()` and `_casefold_parts()` methods.

Makes the code a little more direct.

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

barneygale committed:
Further simplify overall diff by inlining code from `compile_pattern()`

brettcannon commented on Lib/pathlib.py:
```diff
@@ -23,6 +23,14 @@
 # Internals
 #
 
+_WIN_EXT_NAMESPACE_PREFIX = '\\\\?\\'
```
What is this supposed to be? Is this eventually used as a regex? And could you use a raw string to make it a bit more readable?

arhadthedev replied:
> And could you use a raw string to make it a bit more readable?

I wanted to propose the same, but unfortunately no:

```python
>>> r'\\?\'
  File "<stdin>", line 1
    r'\\?\'
    ^
SyntaxError: unterminated string literal (detected at line 1)
>>> r'\\?\\'
'\\\\?\\\\'
```

barneygale replied:
This was previously defined as `_WindowsFlavour.ext_namespace_prefix`. It's used when parsing UNC paths in `PurePath._split_extended_path()`, which is called by `PurePath._splitroot()`, which is a method similar to `os.path.splitdrive()`. It's only used once, so I'll replace its usage with a string literal.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -23,6 +23,14 @@
 #
 
+_WIN_EXT_NAMESPACE_PREFIX = '\\\\?\\'
+_WIN_DRIVE_LETTERS = set('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ')
```
I assume this is meant to be immutable?

```suggestion
_WIN_DRIVE_LETTERS = frozenset(string.ascii_letters)
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -347,7 +126,8 @@ def _select_from(self, parent_path, is_dir, exists, scandir):
     def __init__(self, pat, child_parts, flavour):
-        self.match = flavour.compile_pattern(pat)
+        flags = 0 if flavour is posixpath else re.IGNORECASE
+        self.match = re.compile(fnmatch.translate(pat), flags).fullmatch
```
Since the default value is likely never going to change this is fine, but you could also do:
```suggestion
        args = {}
        if flavour is not posixpath:
            args["flags"] = re.IGNORECASE
        self.match = re.compile(fnmatch.translate(pat), **args).fullmatch
```

barneygale replied:
Hm. There's also `re.NOFLAG` as of fea7290a0ec - might be best of both worlds? Will pop a commit in.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+        prefix = ''
+        if s.startswith(_WIN_EXT_NAMESPACE_PREFIX):
+            prefix = s[:4]
+            s = s[4:]
```
Is `_WIN_EXT_NAMESPACE_PREFIX` 4 characters long and that's why you are slicing that much? If so:
```suggestion
            prefix = _WIN_EXT_NAMESPACE_PREFIX
            s = s.removesuffix(_WIN_EXT_NAMESPACE_PREFIX)
```

Regardless, the length of 3 seems a bit magical.

barneygale replied:
I'd like to leave this code alone (aside from moving it) as it'll be removed in #91882 - see my top-level comment. Hope that makes sense! Resolving this thread...

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+        if s.startswith(_WIN_EXT_NAMESPACE_PREFIX):
+            prefix = s[:4]
+            s = s[4:]
+            if s.startswith('UNC\\'):
```
```suggestion
            if s.startswith(r'UNC\'):
```

barneygale replied:
```
SyntaxError: unterminated string literal
```

Gotta love the r-string gotchas :D

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            s = s[4:]
+            if s.startswith('UNC\\'):
+                prefix += s[:3]
+                s = '\\' + s[3:]
```
```suggestion
                extra_prefix, sep, rest = s.partition(r'\')
                prefix += extra_prefix
                s = sep + rest
```

barneygale replied:
I'd like to leave this code alone (aside from moving it) as it'll be removed in #91882 - see my top-level comment. Hope that makes sense! Resolving this thread...

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            else:
+                return '', '', part
+
+        first = part[0:1]
```
```suggestion
        first = part[0]
```

brettcannon replied:
`first` and `second` of what? The single slice isn't descriptive.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+
+        first = part[0:1]
+        second = part[1:2]
+        if (second == sep and first == sep):
```
```suggestion
        if second == sep and first == sep:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+        first = part[0:1]
+        second = part[1:2]
+        if (second == sep and first == sep):
+            # XXX extended paths should also disable the collapsing of "."
```
There shouldn't be an `XXX` comments left.

barneygale replied:
Is that true even when moving code around? Fixing other bugs increases the overall risk of the patch.

eryksun replied:
The comment is mistaken. The Windows API skips normalizing extended paths only when they're opened. `GetFullPathNameW()`, on the other hand, normalizes all UNC paths, including "\\\\.\\" and "\\\\?\\" paths. This includes stripping trailing dots and spaces from the final path component, which has yet to be implemented by either `ntpath.normpath()` or `pathlib`.

Currently `pathlib` doesn't support literal paths that skip normalization, such as allowing "." components and repeated separators. It doesn't even support an initial "." component such as "./spam", which is sometimes necessary in both POSIX and Windows.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            # XXX extended paths should also disable the collapsing of "."
+            # components (according to MSDN docs).
+            prefix, part = cls._split_extended_path(part)
+            first = part[0:1]
```
```suggestion
            first = part[0]
```

barneygale replied:
Raises `IndexError` if `part` is zero-length.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+        else:
+            prefix = ''
+        third = part[2:3]
+        if (second == sep and first == sep and third != sep):
```
```suggestion
        if second == sep and first == sep and third != sep:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            prefix = ''
+        third = part[2:3]
+        if (second == sep and first == sep and third != sep):
+            # is a UNC path:
```
What "is a UNC path"? If that's the start of a sentence it should be capitalized.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            # \\machine\mountpoint\directory\etc\...
+            #            directory ^^^^^^^^^^^^^^
+            index = part.find(sep, 2)
+            if index != -1:
```
```suggestion
            if (index := part.find(sep, 2)) != -1:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+                        return part[:index2], sep, part[index2 + 1:]
+        drv = root = ''
+        if second == ':' and first in _WIN_DRIVE_LETTERS:
+            drv = part[:2]
```
What's the `2` based on?

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            second = part[1:2]
+        else:
+            prefix = ''
+        third = part[2:3]
```
What are `first`, `second`, and `third`?

barneygale replied:
Their meaning has not changed in this patch. They are the first, second and third characters of `part`,

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+                return '', '', part
+
+        first = part[0:1]
+        second = part[1:2]
```
```suggestion
        second = part[1]
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            # components (according to MSDN docs).
+            prefix, part = cls._split_extended_path(part)
+            first = part[0:1]
+            second = part[1:2]
```
```suggestion
            second = part[1]
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+            second = part[1:2]
+        else:
+            prefix = ''
+        third = part[2:3]
```
```suggestion
        third = part[2]
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+                index2 = part.find(sep, index + 1)
+                # a UNC path can't have two slashes in a row
+                # (after the initial two)
+                if index2 != index + 1:
```
```suggestion
                # A UNC path can't have two slashes in a row
                # (after the initial two)
                if (index2 := part.find(sep, index + 1)) != index + 1:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+                if index2 != index + 1:
+                    if index2 == -1:
+                        index2 = len(part)
+                    if prefix:
```
```suggestion
                    if index2 == -1:
                        index2 = len(part)
                        
                    if prefix:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -479,6 +260,113 @@ def __reduce__(self):
+        it = reversed(parts)
+        for part in it:
+            if not part:
+                continue
```
```suggestion
        for part in filter(None, reversed(parts)):
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -521,14 +409,31 @@ def _from_parsed_parts(cls, drv, root, parts):
+
+    def _join_parsed_parts(self, drv2, root2, parts2):
+        """
+        Join the two paths represented by the respective
```
```suggestion
        """Join the two paths represented by the respective
```

Could you also make the file consistent in how it formats docstrings?

brettcannon commented on Lib/pathlib.py:
```diff
@@ -562,15 +467,31 @@ def as_uri(self):
             raise ValueError("relative path can't be expressed as a file URI")
-        return self._flavour.make_uri(self)
+
+        # FIXME: move this implementation to os.path.fileuri()
```
Shouldn't have `FIXME` comments left.

barneygale replied:
In what circumstances can TODOs/FIXMEs/etc be added to the code? Are there none?

merwok replied:
Maybe when adding a note to existing code that’s not touched by the PR.  But adding a fixme in new code seems strange, the new code should have the fix! 🙂

brettcannon commented on Lib/pathlib.py:
```diff
@@ -792,20 +714,32 @@ def is_absolute(self):
+        if self._parts[0].startswith('\\\\'):
+            # UNC paths are never reserved
+            return False
+        name = self._parts[-1].partition('.')[0].partition(':')[0].rstrip(' ')
```
I don't think you need the first partition since you are still only operating on the first `:`. Or are you relying on `IndexError` being thrown?

```suggestion
        name = self._parts[-1].partition(':')[0].rstrip(' ')
```

barneygale replied:
These aren't equivalent:

```
>>> 'a.b'.partition('.')[0].partition(':')[0].rstrip(' ')
'a'
>>> 'a.b'.partition(':')[0].rstrip(' ')
'a.b'
```

brettcannon reviewed (changes requested):
Most of the comments are either code touch-ups or asking to make the code more self-describing.

barneygale committed:
Apply suggestions from code review

Co-authored-by: Brett Cannon <brett@python.org>

barneygale committed:
Address a couple bits of review feedback

barneygale committed:
Fix tests

barneygale committed:
Fix docstring formatting.

merwok commented on Lib/pathlib.py:
```diff
@@ -347,7 +126,8 @@ def _select_from(self, parent_path, is_dir, exists, scandir):
     def __init__(self, pat, child_parts, flavour):
-        self.match = flavour.compile_pattern(pat)
+        flags = re.NOFLAG if flavour is posixpath else re.IGNORECASE
+        self.match = re.compile(fnmatch.translate(pat), flags).fullmatch
```
Isn’t case sensitivity a property of each filesystem, not the OS?

barneygale replied:
🤔 yes, but Python (including `os.path` and `pathlib`) makes a hard division between Windows (where we normalise case before making comparisons etc) and POSIX (inc. Mac). Here I've preserved existing behaviour. But @sogom has already found at least one instance where pathlib gets it wrong: #86824

eryksun replied:
@barneygale, NTFS in Windows 10 allows enabling case sensitivity for directory trees, so it's no longer strictly a property of either the filesystem or the OS. Using `ntpath.normcase()`, among other case-insensitive comparisons, is a strategy that's increasingly running into problems. For some cases, such as wildcard matching in a directory listing (when not using `FindFirstFileW()` for a DOS-style wildcard match), one should query whether the listed directory is case sensitive via `GetFileInformationByHandleEx()`: `FileCaseSensitiveInfo` (23). In other cases, such as remembering seen symlink targets, the true normalized path is needed. Call `CreateFileW()` to open the path with reparsing disabled (i.e. `FILE_FLAG_OPEN_REPARSE_POINT`), and then call `GetFinalPathNameByHandleW()` to get the final normalized path.

barneygale replied:
Thank you Eryk. I'd like to log a follow-up bug so we don't lose track of this info, but it's a bit beyond my current level of Windows expertise. It feels like too large a job to solve here. @merwok how should we proceed do you think?

eryksun replied:
Nowadays, Linux EXT4 can be formatted to enable [per-directory case folding](https://www.collabora.com/news-and-blog/blog/2020/08/27/using-the-linux-kernel-case-insensitive-feature-in-ext4/), similar to how this works with NTFS in Windows. The assumption about case sensitivity in POSIX and Windows is wide spread throughout the standard library and import system (e.g. `sys.path`). Supporting per-directory case sensitivity would be a lot of work. I don't know whether it's worth it, but I just wanted to point out the increasing set of cases where our assumption breaks down.

merwok replied:
I see!  OK to keep applying the current division then.  Could someone check if the docs are clear about this assumption?

barneygale replied:
*Fairly* clear IMO. In `os.path`:

![image](https://user-images.githubusercontent.com/960340/166457363-4f98d891-ffa2-497a-ba43-7cf62440ef17.png)

![image](https://user-images.githubusercontent.com/960340/166457427-edd6d0ca-36fa-4902-96e8-5419c69e3a8a.png)

And in `pathlib`:

![image](https://user-images.githubusercontent.com/960340/166457656-6a9067ca-2338-4468-aee0-a4c07f1289d8.png)

barneygale committed:
Remove FIXME comments

gst commented on Lib/pathlib.py:
```diff
@@ -562,15 +465,30 @@ def as_uri(self):
             raise ValueError("relative path can't be expressed as a file URI")
-        return self._flavour.make_uri(self)
+
+        if self._flavour is posixpath:
```
should not use `issubclass` ?

barneygale replied:
`posixpath` is a module, not a class. It serves as the implementation for `os.path` for non-Windows OSs. See the second note in the [`os.path` docs](https://docs.python.org/3/library/os.path.html) for more info.

Unless you're suggesting that `issubclass` works with modules? From a tiny bit of testing that doesn't seem to be the case, but perhaps I'm missing something.

barneygale replied:
Background: pathlib began as a high-level *re-implementation* of `posixpath` and `ntpath`. This ceased to make sense when pathlib was merged into Python itself - why would we have two similar implementations of the same functionality in the same codebase? For the past few Python releases I've been merging these implementations, e.g. pathlib no longer re-implements `os.path.expanduser()` or `os.path.realpath()` -- instead, it wraps them. The result is that pathlib is becoming a high-level *wrapper* of `posixpath` and `ntpath`.

These `flavour is posixpath` expressions are the last vestiges of the pathlib-specific functionality that hasn't been merged back to `posixpath` and `ntpath`. If we want to remove them, we could introduce three new `os.path` functions and call through to them:

- `os.path.fileuri()` -- called by `pathlib.PurePath.as_uri()`
- `os.path.isreserved()` -- called by `pathlib.PurePath.is_reserved()`
- `os.path.splitroot()` -- called by `pathlib.PurePath._parse_args()`

Personally I'm on the fence about these. It seems a bit OTT to introduce new `os.path` functions just to tighten up the pathlib codebase.

gst replied:
my bad @barneygale I had in mind that posixpath was a class here.

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

AlexWaygood commented on Lib/pathlib.py:
```diff
@@ -23,6 +24,13 @@
 #
 
+_WIN_DRIVE_LETTERS = frozenset(string.ascii_letters)
+_WIN_RESERVED_NAMES = (
```
Any reason for this not to be a frozenset too?

```suggestion
_WIN_RESERVED_NAMES = frozenset(
```

barneygale committed:
Update Lib/pathlib.py

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

merwok commented on Lib/pathlib.py:
```diff
@@ -562,15 +465,30 @@ def as_uri(self):
 
     @property
     def _cparts(self):
         # Cached casefolded parts, for hashing and comparison
```
There are multiple comments and variable / attribute names like `_cparts` or `cf` where `c` stands for casefold, they will be harder to understand for new readers.

barneygale committed:
Remove vestigal 'casefold' names.

merwok commented on Lib/test/test_pathlib.py:
```diff
@@ -2785,6 +2776,16 @@ def check():
 
+class PurePathSubclassTest(_BasePurePathTest, unittest.TestCase):
+    class cls(pathlib.PurePath):
+        pass
```
Shouldn’t this be inside a method, to make sure success/failure/error happens properly inside a test case, not when the module is loaded?

barneygale replied:
The `_BasePurePathTest` superclass provides the test methods that use `cls`, e.g.:

```python
    def test_glob(self):
        P = self.cls
        p = P(BASE)
        given = set(p.glob("FILEa"))
        expect = set() if not os_helper.fs_is_case_insensitive(BASE) else given
        self.assertEqual(given, expect)
        self.assertEqual(set(p.glob("FILEa*")), set())
```

The snippet you highlighted sets `cls` for use in these tests. It isn't in itself a test!

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

barneygale committed:
Restore comment on POSIX paths beginning `//` and tweak implementation.

barneygale committed:
Inline `_join_parsed_parts()` and `_make_child()` in `joinpath()`.

AA-Turner commented on Lib/pathlib.py:
```diff
@@ -23,6 +24,12 @@
+_WIN_RESERVED_NAMES = frozenset(
+    {'CON', 'PRN', 'AUX', 'NUL', 'CONIN$', 'CONOUT$'} |
+    {'COM%s' % c for c in '123456789\xb9\xb2\xb3'} |
+    {'LPT%s' % c for c in '123456789\xb9\xb2\xb3'}
```
Might make a tiny difference on module import speed, f-strings are faster
```suggestion
    {f'COM{c}' for c in '123456789\xb9\xb2\xb3'} |
    {f'LPT{c}' for c in '123456789\xb9\xb2\xb3'}
```

barneygale committed:
Update Lib/pathlib.py

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

barneygale committed:
Remove unused import

barneygale committed:
Make NEWS entry more precise.

brettcannon commented on Misc/NEWS.d/next/Library/2022-03-05-02-14-09.bpo-24132.W6iORO.rst:
```diff
@@ -0,0 +1,3 @@
+Make :class:`pathlib.PurePath` and :class:`~pathlib.Path` subclassable.
+Previously, attempting to instantiate a subclass resulted in an
+:exc:`AttributeError` being raised. Patch by Barney Gale.
```
```suggestion
Make :class:`pathlib.PurePath` and :class:`~pathlib.Path` subclassable
(private to start). Previously, attempting to instantiate a subclass
resulted in an :exc:`AttributeError` being raised. Patch by Barney Gale.
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -403,8 +241,9 @@ class PurePath(object):
     __slots__ = (
         '_drv', '_root', '_parts',
-        '_str', '_hash', '_pparts', '_cached_cparts',
+        '_str', '_hash', '_pparts', '_cached_ncparts',
```
What are `_pparts` (path parts) and `_cached_ncparts` (UNC is the best I can think of)? Maybe a comment to explain what the attributes are expected to be used for?

barneygale replied:
`cparts` meant "casefolded parts", and `ncparts` means "normcase()d parts". I'll add a comment.

barneygale replied:
I've renamed them as follows:

- `_pparts` --> `_parts_tuple`
- `_ncparts` --> `_parts_normcase`
- `_cached_ncparts` --> `_parts_normcase_cached`

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+        for part in it:
+            if not part:
+                continue
+            if altsep:
```
```suggestion
            elif altsep:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+                root, rel = sep * 2, rel[2:]
+            elif rel[:1] == sep or drv[:1] == sep:
+                root, rel = sep, rel.lstrip(sep)
+            if sep in rel:
```
If the fall-through is on purpose then a newline helps make that a bit more obvious.

```suggestion

            if sep in rel:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+            else:
+                if rel and rel != '.':
+                    parsed.append(sys.intern(rel))
+            if drv or root:
```
```suggestion

            if drv or root:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+                        drv = cls._flavour.splitdrive(part)[0]
+                        if drv:
+                            break
+                break
```
```suggestion

                break
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+                    for part in it:
+                        if not part:
+                            continue
+                        if altsep:
```
```suggestion
                        elif altsep:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -421,6 +260,54 @@ def __reduce__(self):
+            # "A pathname that begins with two successive slashes may be
+            # interpreted in an implementation-defined manner, although more
+            # than two leading slashes shall be treated as a single slash".
+            if cls._flavour is posixpath and rel[:2] == sep * 2 and rel[2:3] != sep:
```
Is it dangerous to hard-code this logic based on the exact flavour? It's the only piece that's isn't abstracted in this method around `cls._flavour`.

barneygale replied:
This is the trickiest bit of the whole PR I think. 

1. We could use `cls._flavour.normpath()` to normalise the root, but it also collapses `../` segments, and we don't want that behaviour in pathlib. We'd need to stitch together the correct path to make it work.
2. We could add a *keep_pardir* argument to `normpath()` (see #88482), but there's a C implementation that would also need adjustment, and it feels like a can of worms.
3. We could add something new to `os.path` like `normroot()`, or perhaps `root` and `altroot`. But they'd be pretty niche!

There are probably options I'm missing. Not sure of the best way forward here - what do you reckon?

barneygale replied:
I've made `_parse_parts()` call `normpath()` only on the root, which seems to do the trick :-)

brettcannon commented on Lib/pathlib.py:
```diff
@@ -504,48 +385,63 @@ def as_uri(self):
-    def _cparts(self):
-        # Cached casefolded parts, for hashing and comparison
+    def _ncparts(self):
+        # Cached normcased parts, for hashing and comparison
```
```suggestion
        # Cached normcased parts, for hashing and comparison.
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -664,8 +560,9 @@ def relative_to(self, *other):
         n = len(to_abs_parts)
-        cf = self._flavour.casefold_parts
-        if (root or drv) if n == 0 else cf(abs_parts[:n]) != cf(to_abs_parts):
+        def nc(parts):
```
Why the function defined here? I would either abstract it out, move it to the top with some whitespace, or better yet just expand the code so it isn't so compact as that ternary operator within an `if` guard is really hard to follow (too many `if` in a single line).

brettcannon commented on Lib/pathlib.py:
```diff
@@ -664,8 +560,9 @@ def relative_to(self, *other):
-        if (root or drv) if n == 0 else cf(abs_parts[:n]) != cf(to_abs_parts):
+        def nc(parts):
+            return [self._flavour.normcase(p) for p in parts]
+        if (root or drv) if n == 0 else nc(abs_parts[:n]) != nc(to_abs_parts):
```
See commend above about `nc()` to decide whether to take this or do something else.

```suggestion
        if not n:
            is_error = root or drive
        else:
            nc_abs_parts = list(map(self._flavour.normcase, abs_parts[:n]))
            nc_to_abs_parts = list(map(self._flavour.normcase, to_abs_parts[:n]))
            is_error = ns_abs_parts != nc_to_abs_parts
        if is_error:
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -734,27 +643,38 @@ def is_absolute(self):
         if not self._root:
             return False
-        return not self._flavour.has_drv or bool(self._drv)
+        return self._flavour is posixpath or bool(self._drv)
```
Is it good to hard-code a check for a specific flavour?

barneygale replied:
I'd like to implement this via:

```python
return self._flavour.isabs(self)
```

... but `ntpath.isabs()` doesn't agree with `PureWindowsPath.is_absolute()` (see #44626) and there's no realistic prospect of fixing it, as far as I can tell.

Perhaps this would be better?

```python
# ntpath.isabs() is defective
if self._flavour is ntpath:  
    return bool(self._root and self._drv)

return self._flavour.isabs(self)
```

eryksun replied:
UNC paths are always absolute, i.e. are never resolved against the current working directory, even if they're incomplete or invalid and have no drive, such as "//", "///", "//spam", or "//?/UNC/spam". The initial parsing could set a `_unc` attribute. Then for the `ntpath` flavour, `is_absolute()` could use  `bool(self._unc or self._root and self._drv)`.

barneygale replied:
`_parse_parts()` already handles this:

```python
elif rel[:1] == sep or drv[:1] == sep:
    root, rel = sep, rel.lstrip(sep)
```

If the drive is a UNC path, we set the root to `'\\'`.

eryksun replied:
Setting `root` to a single backslash has nothing to do with this. The problem is that these cases have *no drive*, but they're still absolute paths.

barneygale replied:
Right. I'm not sure what to say except that I'm not planning to fix that in this PR. It's not a regression and it's not directly related to the aim of the PR, unless I'm still misunderstanding you.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -734,27 +643,38 @@ def is_absolute(self):
         """Return True if the path contains one of the special names reserved
         by the system, if any."""
-        return self._flavour.is_reserved(self._parts)
+        if self._flavour is posixpath or not self._parts:
```
Is it good to hard-code a check for a specific flavour?

barneygale replied:
I'd like to do this

```python
return self._flavour.isreserved(self)
```

But it would require a new `os.path.isreserved()` function - see #88569. What do you think?

barneygale replied:
I've created a PR that adds `os.path.isreserved()`: #95486.

barneygale replied:
@brettcannon FYI, Eryk Sun and Steve Dower are helping refine the `os.path.isreserved()` implementation in #95486. It will take a while to get right! As such I'm going to push a commit to _this_ PR that adds `os.path._isreserved()` (note the underscore). Hopefully unblocks this PR! I'll fix up merge conflicts from `main` when they happen.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -734,27 +643,38 @@ def is_absolute(self):
-        cf = self._flavour.casefold
-        path_pattern = cf(path_pattern)
-        drv, root, pat_parts = self._flavour.parse_parts((path_pattern,))
+        nc = self._flavour.normcase
```
You have now defined this twice with the same, but in two different ways. It might be best to make a method.

brettcannon commented on Lib/pathlib.py:
```diff
@@ -734,27 +643,38 @@ def is_absolute(self):
         if not pat_parts:
             raise ValueError("empty pattern")
-        if drv and drv != cf(self._drv):
+        if drv and drv != nc(self._drv):
```
```suggestion
        elif drv and drv != nc(self._drv):
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -734,27 +643,38 @@ def is_absolute(self):
+        if drv and drv != nc(self._drv):
             return False
-        if root and root != cf(self._root):
+        if root and root != nc(self._root):
```
```suggestion
        elif root and root != nc(self._root):
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -1202,6 +1122,9 @@ def is_mount(self):
         """
         Check if this path is a POSIX mount point
         """
+        if self._flavour is not posixpath:
```
You know what I'm going to ask about this line. 😉

barneygale replied:
I'd like to make this:

```
return self._flavour.ismount(self)
```

See also #86943.

However, the current implementation will be useful for `AbstractPath`, so I've been avoiding deleting it. Perhaps I should make the change now, and rescue the old implementation from Git history when we come to add `AbstractPath`?

barneygale replied:
I've marked #31458 as ready for review

brettcannon replied:
Well, I just merged #31458, so now you can make the change. 🙂

barneygale replied:
\o/ I've merged in main.

brettcannon reviewed (changes requested):
Overall I like the simplicity that a class attribute that models `os.path` is all that's needed to make a subclass work. To that end, though, I have asked in every place you hard-coded a check against `posixpath` if that's the best option. It's totally acceptable to use the `os.path` trick as a short cut, but then to also have other methods or class attributes to potentially override in cases where what to do isn't necessarily clear based on info gleaned from `os.path`. In other words, every place you hard-code a check for `posixpath` is one place where subclassing is potentially broken (and don't assume we know best; people do some weird stuff when it comes to interfaces 😅).

Otherwise some stylistic points and making the news entry clear this is not ready for public consumption (yet).

barneygale committed:
Apply suggestions from code review

Co-authored-by: Brett Cannon <brett@python.org>

barneygale committed:
Clarify `_pparts` and `_ncparts` naming.

- `_pparts` --> `_parts_tuple`
- `_ncparts` --> `_parts_normcase`
- `_cached_ncparts` --> `_parts_normcase_cached`

barneygale committed:
Tidy up usage of `normcase()`

eryksun commented on Lib/pathlib.py:
```diff
@@ -862,7 +788,7 @@ def samefile(self, other_path):
         except AttributeError:
             other_st = self.__class__(other_path).stat()
-        return os.path.samestat(st, other_st)
+        return self._flavour.samestat(st, other_st)
```
Basing `samefile()` on `stat()` is problematic in Windows. The volume serial number is used for `st_dev`. The VSN of a filesystem can be any 32-bit value. It isn't forced to be unique, but usually it's a random value that's likely to be unique. The file ID is used for `st_ino`. It's 0 if a filesystem doesn't support file IDs. WebDAV is a common example that always returns 0 for both the VSN and file IDs, so the `stat()` result is completely worthless for `samefile()` within or between WebDAV filesystems. This isn't buggy behavior. The WebDAV filesystem in Windows is completely within its rights to behave this way according to documented protocols.

This method should call `self._flavour.samefile()` instead of `samestat()`. It currently has the same design flaw, but if we're going to fix this anywhere, it's by moving the Windows implementations of `samefile()` and `sameopenfile()` back into `ntpath`. There are open issues for this, but there's no momentum pushing for a resolution.

barneygale replied:
I agree, but I don't think this is the right place to fix it. Everything I add to this PR makes it harder to review and less likely to land.

eryksun replied:
When implementing `samefile()`, I should think it would be an easy sell to use `self._flavour.samefile()`. This is more flexible for subclasses since it doesn't assume any `stat()` semantics that may not be supported. The fact that `st_dev` and `st_ino` are unreliable in Windows is just one example.

barneygale committed:
Tweak `is_absolute()` to use `os.path.isabs()` wherever possible.

barneygale committed:
Stop using `self._flavour` in `as_uri()`.

barneygale committed:
Remove `cls._flavour` identity check in `_parse_parts()`.

barneygale committed:
Fix Windows tests

barneygale committed:
Restore `_split_root()` method and its tests.

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

barneygale committed:
Optimize `_split_root()`

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

barneygale commented on Lib/pathlib.py:
```diff
@@ -407,6 +271,51 @@ def __reduce__(self):
         return (self.__class__, tuple(self._parts))
 
+    @classmethod
+    def _split_root(cls, part):
```
I plan to tidy this method up and move its implementation into `ntpath` and `posixpath`, but I'd like to do it in a subsequent PR that allows users to swap in their own path normalization logic.

eryksun replied:
Is calling `normpath()` and `splitdrive()` on the anchor just to support a "//" root on POSIX? I agree that this should use a new `splitroot()` function, implemented separately in `ntpath` and `posixpath`.

barneygale replied:
It's used to normalize roots on both POSIX (like `///etc` -> `/etc`) and Windows (like `C:\\foo` -> `C:\foo`). This could be implemented more efficiently (and obviously) with dedicated functions in `ntpath` and `posixpath`. I think I'll have a PR for that within the next six months.

barneygale committed:
Move `is_reserved()` implementation into `os.path` as a private function.

barneygale committed:
Use `os.path.normcase()` when globbing.

This removes a `flavour is posixpath` check in `_WildcardSelector`.
Instead, we normalise the case of both the pattern and the `os.scandir()`
entries, and then perform a case-sensitive comparison. The same approach is
taken by `fnmatch.fnmatch()`.

barneygale committed:
Simplify `_split_root()` implementation.

barneygale committed:
Remove tests for malformed UNC paths

brettcannon committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

brettcannon commented on Lib/pathlib.py:
```diff
@@ -407,6 +265,33 @@ def __reduce__(self):
+    def _split_root(cls, part):
+        sep = cls._flavour.sep
+        rel = cls._flavour.splitdrive(part)[1].lstrip(sep)
+        anchor = part[:len(part)-len(rel)]
```
```suggestion
        anchor = part[:-len(rel)]
```

barneygale replied:
Not equivalent when `rel` is empty, as you get `part[:-0]`, which is `part[:0]`

barneygale replied:
Just realised I can use `removesuffix()` here, d'oh.

brettcannon commented on Lib/ntpath.py:
```diff
@@ -311,6 +311,30 @@ def ismount(path):
 
+# Is the path reserved?
+# This private function is used by pathlib.PureWindowsPath.is_reserved().
+# See GH-88569 for a proposal to make this function public.
```
I think we should keep this in `pathlib` until/unless it is decided to take it public. Otherwise having pathlib leak into `os.path` just for purity of an internal API seems unnecessary.

brettcannon reviewed (changes requested):
Only real change is to not put code into `os.path` just for the sake of internal API purity.

barneygale committed:
Revert "Move `is_reserved()` implementation into `os.path` as a private function."

This reverts commit 806ad150ac5f7397623ea5d0c6869f3b7d43b6a1.

barneygale committed:
Tweak _split_root() to use removesuffix()

barneygale committed:
Remove test that `/b/c/d` and `///b/c/d` are equivalent.

This is no longer true as of 99fcf15052, see #96290

barneygale committed:
Remove test that `/b/c/d` and `///b/c/d` are equivalent.

This is no longer true as of 99fcf15052, see #96290

merwok commented on Lib/pathlib.py:
```diff
@@ -52,150 +58,6 @@ def _is_wildcard_pattern(pat):
-
-class _WindowsFlavour(_Flavour):
-    # Reference for Windows paths can be found at
-    # http://msdn.microsoft.com/en-us/library/aa365247%28v=vs.85%29.aspx
```
What do you think of keeping this reference in the new location?

barneygale replied:
I think that would be near the top of `ntpath.py` - does that work for you?

merwok replied:
For sure, above _WIN_RESERVED_NAMES

barneygale committed:
Restore comment linking to *Naming Files* page in Microsoft docs

brettcannon commented on Lib/pathlib.py:
```diff
@@ -30,6 +30,14 @@
 #
 
+# Reference for Windows paths can be found at
+# https://docs.microsoft.com/en-gb/windows/win32/fileio/naming-a-file
```
```suggestion
# https://docs.microsoft.com/en-gb/windows/win32/fileio/naming-a-file .
```

eryksun replied:
```suggestion
# https://learn.microsoft.com/en-gb/windows/win32/fileio/naming-a-file .
```

"docs\.microsoft\.com" has recently migrated to "learn.microsoft.com". 

Note that omitting the language from the URL (e.g. "en-gb") would automatically redirect to the user's preferred language (e.g. "learn\.microsoft\.com/ru-ru/" or "learn\.microsoft\.com/en-us/").

brettcannon commented on Lib/pathlib.py:
```diff
@@ -675,22 +581,34 @@ def parts(self):
+                # Same drive => second path is relative to the first
+                return self._from_parsed_parts(drv1, root1, parts1 + parts2[1:])
+        else:
+            # Second path is non-anchored (common case)
```
```suggestion
            # Second path is non-anchored (common case).
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -675,22 +581,34 @@ def parts(self):
-        return self._make_child(args)
+        drv1, root1, parts1 = self._drv, self._root, self._parts
+        drv2, root2, parts2 = self._parse_args(args)
+        if root2:
+            if not drv2 and drv1:
+                return self._from_parsed_parts(drv1, root2, [drv1 + root2] + parts2[1:])
```
```suggestion
        if root2:
            if not drv2 and drv1:
                return self._from_parsed_parts(drv1, root2, [drv1 + root2] + parts2[1:])
            else:
                return self._from_parsed_parts(drv2, root2, parts2)
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -675,22 +581,34 @@ def parts(self):
+        elif drv2:
+            if drv2 == drv1 or self._flavour.normcase(drv2) == self._flavour.normcase(drv1):
+                # Same drive => second path is relative to the first
+                return self._from_parsed_parts(drv1, root1, parts1 + parts2[1:])
```
```suggestion
        elif drv2:
            if drv2 == drv1 or self._flavour.normcase(drv2) == self._flavour.normcase(drv1):
                # Same drive => second path is relative to the first.
                return self._from_parsed_parts(drv1, root1, parts1 + parts2[1:])
            else:
                return self._from_parsed_parts(drv2, root2, parts2)
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -675,22 +581,34 @@ def parts(self):
+        else:
+            # Second path is non-anchored (common case)
+            return self._from_parsed_parts(drv1, root1, parts1 + parts2)
+        return self._from_parsed_parts(drv2, root2, parts2)
```
Having an `else` clause with a `return` statement preceding this makes you do a double-take on what's going on and how you may end up at this line. As such, might as well inline the lines above for logical simplicity rather than avoiding writing the same line twice.

```suggestion
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -718,29 +636,40 @@ def parents(self):
-        if not self._root:
-            return False
-        return not self._flavour.has_drv or bool(self._drv)
+        # ntpath.isabs() is defective - see GH-44626
```
```suggestion
        # ntpath.isabs() is defective - see GH-44626 .
```

brettcannon commented on Lib/pathlib.py:
```diff
@@ -718,29 +636,40 @@ def parents(self):
+        # exist). We err on the side of caution and return True for paths
+        # which are not considered reserved by Windows.
+        if self._parts[0].startswith('\\\\'):
+            # UNC paths are never reserved
```
```suggestion
            # UNC paths are never reserved.
```

brettcannon reviewed (changes requested):
Minor things to change while we wait for the PR to handle the merge conflicts. Otherwise it LGTM!

barneygale committed:
Apply suggestions from code review

Co-authored-by: Eryk Sun <eryksun@gmail.com>
Co-authored-by: Brett Cannon <brett@python.org>

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

barneygale committed:
Clarify repr roundtrip tests.

barneygale committed:
Reduce tests diff

barneygale committed:
Fix missing full stops.

barneygale committed:
Remove four more test assertions affected by #96290

brettcannon reviewed (changes requested):
Merge conflicts

barneygale committed:
Merge branch 'main' into bpo-44136-remove-pathlib-flavour-2

brettcannon commented on Lib/test/test_pathlib.py:
```diff
@@ -182,8 +177,7 @@ class _BasePurePathTest(object):
             ('', 'a', 'b'), ('a', '', 'b'), ('a', 'b', ''),
             ],
         '/b/c/d': [
-            ('a', '/b/c', 'd'), ('a', '///b//c', 'd/'),
```
Is there a specific reason for dropping the `///b//c` test?

eryksun replied:
"///b//c" is an invalid UNC path. The server name is empty, and the share name is "b". It isn't equivalent to "/b/c". 

WinAPI `GetFullPathNameW()`:

```pycon
>>> os.getcwd()
'C:\\Temp'

>>> nt._getfullpathname('///b/c/../..')
'\\\\\\b'
>>> nt._getfullpathname('/b/c/../..')
'C:\\'
```

WinAPI `PathCchSkipRoot()`, the basis for `nt._path_splitroot()`:

```pycon
>>> nt._path_splitroot('///b/c')
('///b/', 'c')
>>> nt._path_splitroot('/b/c')
('/', 'b/c')

```

Since the current behavior of pathlib is wrong, I suppose Barney is dropping the test for now until the correct behavior can be tested. There are two options to fix this, if I recall correctly. Either `ntpath.splitdrive()` can be changed to split "///b" as a drive, or pathlib can handle such cases specially.

barneygale replied:
I found that calling `ntpath.normpath()` on some invalid UNC paths produced different results depending on whether or not Python is running on a Windows system. I logged this here: #96290.

That bug seems to be fairly minor; I'm the only person to report it, months after it was introduced. The expected behaviour isn't clear-cut.

In this PR, we use `normpath()` to normalize path anchors, which means we're newly affected by the bug. This causes a pathlib test failure, but as the expected behaviour isn't clear, I figured that removing the test is a reasonable way forward. Happy to reconsider!

barneygale replied:
I removed four more assertions (all invalid UNC paths) for the same reason: https://github.com/barneygale/cpython/commit/fbaadf45d443c33a1df8d1122aee0805de3c7822

barneygale replied:
I've proposed a fix (#100351) for #96290, but it's worth noting that it removes the same equivalence test as this PR. As such I don't think it blocks this PR.

brettcannon commented on Lib/test/test_pathlib.py:
```diff
@@ -299,6 +293,11 @@ def test_repr_common(self):
+            r = repr(p)
             # The repr() roundtrips.
             q = eval(r, pathlib.__dict__)
             self.assertIs(q.__class__, p.__class__)
```
Can you put all of this in a `self.subTest()` block?

https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

brettcannon reviewed (changes requested):
One small tweak to the tests, otherwise LGTM!

barneygale committed:
Wrap repr test cases in `self.subTest()` block

eryksun commented on Lib/pathlib.py:
```diff
@@ -405,6 +271,33 @@ def __reduce__(self):
+            anchor = cls._flavour.normpath(anchor)
+        drv, root = cls._flavour.splitdrive(anchor)
+        if drv.startswith(sep):
+            # UNC paths always have a root.
```
Note that device paths that begin with "\\\\.\\" or "\\\\?\\" do not have an implicit root, except for "\\\\?\\UNC\\server\\share".

barneygale replied:
Yep understood. Pathlib wrongly assumes that all UNC paths have a root, but that behaviour has been present from the first versions of pathlib, so I'm not introducing a regression here I don't think.

I will fix this, but it will take some time.

Edit to add some detail: my thinking is that we'll move pathlib towards treating the underlying path as an opaque object. We wouldn't split it up into `(drv, root, parts)`, and so pathlib's "normalization" pass becomes a one-shot string-to-string function, similar to `normpath()`, and probably disableable and user-replaceable. We'll increasingly use `os.path` for low-level operations, rather than re-implementing them in pathlib. Somewhere along the line I'm sure we'll fix pathlib's support for UNC paths by using the right `ntpath` functions in the right way.
